addons/web/static/tests/core/template_inheritance_tests.js

```diff
@@ -0,0 +1,316 @@
+/** @odoo-module **/
+
+import { applyInheritance } from "@web/core/template_inheritance";
+
+const parser = new DOMParser();
+const serializer = new XMLSerializer();
+
+export function _applyInheritance(arch, inherits) {
+    const archXmlDoc = parser.parseFromString(arch, "text/xml");
+    const inheritsDoc = parser.parseFromString(inherits, "text/xml");
+    const modifiedTemplate = applyInheritance(
+        archXmlDoc.documentElement,
+        inheritsDoc.documentElement
+    );
+    return serializer.serializeToString(modifiedTemplate);
+}
+
+QUnit.module("Template Inheritance", {});
+
+QUnit.test("no operation", async (assert) => {
+    const arch = `<t t-name="web.A"> <div><h2>Title</h2>text</div> </t>`;
+    const operations = `<t/>`;
+    assert.strictEqual(_applyInheritance(arch, operations), arch);
+});
+
+QUnit.test("single operation: replace", async (assert) => {
+    const toTest = [
+        {
+            arch: `<t t-name="web.A"> <div><h2>Title</h2>text</div> </t>`,
+            operations: `
+                <t>
+                    <xpath expr="./div/h2" position="replace"><h3>Other title</h3></xpath>
+                </t>`,
+            result: `<t t-name="web.A"> <div><h3>Other title</h3>text</div> </t>`,
+            // TODO check if text should be there? (I think there is a bug in python code)
+        },
+    ];
+    for (const { arch, operations, result } of toTest) {
+        assert.strictEqual(_applyInheritance(arch, operations), result);
+    }
+});
+
+QUnit.test("single operation: replace root (and use a $0)", async (assert) => {
+    const toTest = [
+        {
+            arch: `<t t-name="web.A"> <div>I was petrified</div> </t>`,
+            operations: `
+                <t>
+                    <xpath expr="." position="replace"><div>At first I was afraid</div>$0</xpath>
+                </t>`,
+            result: `<div t-name="web.A">At first I was afraid</div>`,
+            // in outer mode with no parent only first child of operation is kept
+        },
+        {
+            arch: `<t t-name="web.A"> <div>I was petrified</div> </t>`,
+            operations: `
+                <t>
+                    <xpath expr="." position="replace"> <div>$0</div><div>At first I was afraid</div> </xpath>
+                </t>`,
+            result: `<div t-name="web.A"><t t-name="web.A"> <div>I was petrified</div> </t></div>`,
+        },
+        {
+            arch: `<t t-name="web.A"> <div>I was petrified</div> </t>`,
+            operations: `
+                <t>
+                    <xpath expr="." position="replace"> <t><t t-if="cond"><div>At first I was afraid</div></t><t t-else="">$0</t></t> </xpath>
```

Comment:
Does the extra empty `<t>` around the if serve a purpose?

Reply:
Yes. Here the target is the root and the mode outer, so we replace root by root operationContent.cloneNode(true); which is the <t> you mention